cpp/density/SphereVoxelization.cc

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) 2010-2020 The Regents of the University of Michigan
+// This file is from the freud project, released under the BSD 3-Clause License.
+
+#include <cmath>
+#include <stdexcept>
+
+#include "SphereVoxelization.h"
+
+/*! \file SphereVoxelization.cc
+    \brief Routines for computing voxelized densities from spheres centered at points.
+*/
+
+namespace freud { namespace density {
+
+SphereVoxelization::SphereVoxelization(vec3<unsigned int> width, float r_max)
+    : m_width(width), m_r_max(r_max)
+{
+    // dummy initial box, dimensionality of the calculation will be defined by
+    // the first neighbor query passed to the compute method
+    m_box = NULL;
+
+    if (r_max <= 0.0f)
+        throw std::invalid_argument("SphereVoxelization requires r_max to be positive.");
+}
+
+//! Get a reference to the last computed voxels.
+const util::ManagedArray<unsigned int>& SphereVoxelization::getVoxels() const
+{
+    return m_voxels_array;
+}
+
+//! Get width.
+vec3<unsigned int> SphereVoxelization::getWidth() const
+{
+    return m_width;
+}
+
+//! Compute the voxels array.
+void SphereVoxelization::compute(const freud::locality::NeighborQuery* nq)
+{
+    // set the number of dimensions for the calculation the first time it is done
+    if (m_box == NULL)
+        m_box = nq->getBox();
+
+    // Don't allow changes in the number of dimensions for the calculation
+    // after its been set.
+    if (nq->getBox().is2D() != m_box.is2D())
+    {
+        throw std::invalid_argument(
+            "Each SphereVoxelization instance should be used to do calculations "
+            "in a fixed number of dimensions. You have changed the number of "
+            "dimensions used going from one calculation to another.");
+    }
```

Comment:
With my comment above, this would become the following. Note that error messages should avoid using second-person "you" in their text. It sounds less accusatory that way. 😉
```suggestion
    // Set the number of dimensions for the calculation the first time it is done
    if (!m_has_computed || nq->getBox().is2D() != m_box.is2D())
    {
        m_box = nq->getBox();
        m_has_computed = true;
    } else {
        throw std::invalid_argument(
            "The dimensionality of the box passed to SphereVoxelization has changed. "
            "A new instance must be created to handle a different dimensionality.");
    }
```

Reply:
(And remove the second assignment of `m_box` below, as needed.)

Reply:
I think the boolean condition in the if statement here should read `!m_has_computed || nq->getBox().is2D() == m_box.is2D()`

Reply:
Right. Good catch.